### ISSUE TITLE
recursive deep merge / evolve

### ISSUE THREAD

**Issue by ccorcos** [ramda/ramda#1130]
[`merge`](https://github.com/ramda/ramda/blob/master/src/merge.js) and [`evolve`](https://github.com/ramda/ramda/blob/master/src/evolve.js) both use [`_extend`](https://github.com/ramda/ramda/blob/master/src/internal/_extend.js) to do a shallow merge / evolve. 

I think it would be really convenient if these were both deep:

```js
function merge(a, b) {
  return _.extend(clone(a), b);
});

function isPlainObject(value) {
  return Object.prototype.toString.call(value) == '[object Object]';
}

function _extend(destination, other) {
  var props = keys(other);
  var idx = -1, length = props.length;
  while (++idx < length) {
    if (isPlainObject(other[props[idx])) {
      destination[props[idx]] = _extend(destination[props[idx]], other[props[idx]])
    } else {
      destination[props[idx]] = other[props[idx]];
    }
  }
  return destination;
};

function evolve(a, b) {
  return _.evolve(clone(a), b);
});

function isFunction(value) {
  return Object.prototype.toString.call(value) == '[object Function]';
}

function _evolve(destination, other) {
  var props = keys(other);
  var idx = -1, length = props.length;
  while (++idx < length) {
    if (isPlainObject(other[props[idx])) {
      destination[props[idx]] = _evolve(destination[props[idx]], other[props[idx]])
    } else if (isFunction(other[props[idx])) {
      destination[props[idx]] = other[props[idx]](destination[props[idx]]);
    } else {
      destination[props[idx]] = other[props[idx]];
    }
  }
  return destination;
};
```

**Comment by paldepind** [comment]
`R.merge

**Comment by ccorcos** [comment]
Awesome. I like `R.set`. How about making `R.evolve` deep as well?

**Comment by paldepind** [comment]
> How about making R.evolve deep as well?

See:

> I plan to soon do a PR that makes R.evolve recursive/deep just like R.set is.

**Comment by ccorcos** [comment]
sweet :+1: I cant wait.

**Comment by ccorcos** [comment]
I'd also like to suggest that `R.evolve` defaults to the `R.set` functionality when a value isnt a function. For example:

```js
R.evolve({name: 'joe', info: {age: R.add(1)}}, {name:'chet', info:{age:24}})
// => {name: 'joe', info: {age: 25}}
```

**Comment by CrossEye** [comment]
:+1: on updating `evolve` to work deeply.

:-1: on handling non-function properties, for the same reasons I gave in https://github.com/ramda/ramda/issues/1131#issuecomment-106880720

**Comment by paldepind** [comment]
@ccorcos I agree with @CrossEye. You can use `R.always` in such cases.

**Comment by ccorcos** [comment]
true

**Comment by paldepind** [comment]
Closed by #1141.
